pkg/controller/barrier.go

```diff
@@ -0,0 +1,223 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"fmt"
+	"sync"
+)
+
+// ErrorPolicy is used to propagate errors from tasks started by a barrier, and
+// also to signal to the barrier that it should terminate early (i.e. refuse to
+// create more tasks).
+type ErrorPolicy interface {
+	// Apply receives a (possibly nil) error on each completion of a task. It
+	// returns a bool indicating whether the barrier should continue creating
+	// tasks. The barrier will synchronize calls to the ErrorPolicy and make no
+	// more calls once Wait() returns. If state modified by the ErrorPolicy needs
+	// to be accessed outside of the ErrorPolicy but before Wait() returns, it's
+	// the ErrorPolicy's responsibility to synchronize access to this state.
+	Apply(error) bool
+}
+
+// Limiter is used by an task barrier to limit the rate that it spawns tasks.
+// Calls to limiter are synchonized by the taskBarrier.
+type Limiter interface {
+	// reserve attempts reserve a new slot for the task. It returns true if the
+	// slot was successfully reserved.
+	reserve() bool
+
+	// release is called to inform to the limiter that a single task has
+	// completed. The limiter is expected to call Signal() or Broadcast() on the
+	// condition to wake goroutines that may be waiting for a reservation slot.
+	release(*sync.Cond)
+}
+
+// TaskBarrier is in incremental barrier that makes it easy to create multiple
+// async operations and wait for them as a group. It allows creation to be rate
+// limitted by a Limiter. It reports errors to a ErrorPolicy, which aggregates
+// the errors and decides whether the barrier should continue or bail out
+// early. It takes care of error prone requirements, such as remembering to
+// count the start tasks with a waitgroup and synchronizing completion
+// reporting out of the task loop.
+type TaskBarrier struct {
+	// cond guards all state including the policy.
+	cond *sync.Cond
+	// Policy is the ErrorPolicy that the task barrier will apply to errors
+	// returned by tasks.
+	Policy ErrorPolicy
+	// Limiter is the limiter that the task barrier will use to limit the
+	// creation of tasks.
+	Limiter Limiter
```

Comment:
You could just have a nil-able SlowStarter field here, and on nil just skip limiting.

Although, I'm not sure what the benefit is of using TaskBarrier with NoLimiter.
It's basically shooting off a bunch of goroutines, where some may or may not be skipped if ErrorPolicy triggers.
Depending on scheduling order, all tasks may be executed even if ErrorPolicy triggers. So what's the benefit?

Reply:
The idea is that this shows up in godoc nicely but all implementations must be in this package.

Reply:
> Although, I'm not sure what the benefit is of using TaskBarrier with NoLimiter.
> It's basically shooting off a bunch of goroutines, where some may or may not be skipped if ErrorPolicy triggers.

Symmetry between create and delete is what I'm going for. One way to report status out of the async loop.

> Depending on scheduling order, all tasks may be executed even if ErrorPolicy triggers. So what's the benefit?

This is definitely different than the previous behavior. The guarantee provided by the barrier is that no new tasks start after error policy is triggered. Given that the tasks are going to be syscall heavy, and the lock in general won't be contended (entrances inside the barrier are quick, but could happen if the policy is doing way too much work) the overall number of tasks run is unlikely to be significantly affected by scheduling order.

It seems more likely that something inside some tasks block for an extended period while other tasks complete successfully. SlowStarter could be made to match the previous behavior more closely by buffering tokens and releasing them in stages, but the current approach seemed simpler and probably good enough.